### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by niusia-ua**

### CODE REVIEW

**Comment by niusia-ua** [comment]
The `charset` option caused a build error: `error TS5101: Option 'charset' is deprecated and will stop functioning in TypeScript 5.5. Specify compilerOption '"ignoreDeprecations": "5.0"' to silence this error`.

**Comment by niusia-ua** [comment]
I would have liked to add a few upfront micro-optimizations, but they go against your style and make the code a bit unclear 😅

**Comment by KnorpelSenf** [comment]
> The `charset` option caused a build error: `error TS5101: Option 'charset' is deprecated and will stop functioning in TypeScript 5.5. Specify compilerOption '"ignoreDeprecations": "5.0"' to silence this error`.

It had to be removed from all grammY repos, it's good to do this here, too.

**Comment by dcdunkan** [comment]
> I would have liked to add a few upfront micro-optimizations, but they go against your style and make the code a bit unclear 😅

@Nazar-Ant Feel free to add those if it improves the plugin. We'll see if it does affect the style and readability of the code after you add them. And if it's too much, we can revert the changes.

**Comment by niusia-ua** [comment]
In general, it is bad practice to use the `function name() {}` without using `this`. I don't remember exactly, but the reason there is to create a function context, which is a rather difficult operation for a runtime, although modern runtimes can skip this step, but the code has to run for a while to collect statistics. Therefore, it is better to replace all `function name() {}` with variables with arrow functions, which are inherently context-free.

**Comment by niusia-ua** [comment]
By the way, did you know that converting a list of emojis into a JSON object, which is then converted into a JS object (`JSON.parse()`), can speed up the launch of a bot by 1.7x (the exact numbers depend on the runtime and the system itself; I haven't tested this for our emoji list)? The reason is that it's easier for a runtime to parse JSON syntax than JS because of its considerable lightness.

https://v8.dev/blog/cost-of-javascript-2019#json

**Comment by niusia-ua** [comment]
If you still prefer a `function name () {}`, you should place it in a variable, because then the function object will be created lazily instead of instantly, which will delay a runtime-intensive operation. That is:
`function name () {}` - intensive
`const name = function () {}` - simple

**Comment by dcdunkan** [comment]
> In general, it is bad practice to use the `function name() {}` without using `this`. I don't remember exactly, but the reason there is to create a function context, which is a rather difficult operation for a runtime, although modern runtimes can skip this step, but the code has to run for a while to collect statistics. Therefore, it is better to replace all `function name() {}` with variables with arrow functions, which are inherently context-free.

Interesting. I didn't know that. But I'll leave the confirmation to @KnorpelSenf because I don't much about anything :)

Even though generally we tend to use normal function declaration style on the top level across grammY repositories, this change is okay I guess.

**Comment by dcdunkan** [comment]
> By the way, did you know that converting a list of emojis into a JSON object, which is then converted into a JS object (`JSON.parse()`), can speed up the launch of a bot by 1.7x (the exact numbers depend on the runtime and the system itself; I haven't tested this for our emoji list)? The reason is that it's easier for a runtime to parse JSON syntax than JS because of its considerable lightness.
> 
> https://v8.dev/blog/cost-of-javascript-2019#json

Another interesting thing that I didn't know!

Do you think we can use here if it'll improve the performance? But what about auto-completion though? Does that affect it?

**Comment by niusia-ua** [comment]
I think there will be problems with the automatic creation of types as it is now. You will have to create a new huge type for this.

You don't really have to worry about that. It won't play a significant role. Except that it will save a little bit of resources on serverless

**Comment by niusia-ua** [comment]
I'm just bragging about my knowledge. Recently, I came across a Ukrainian programmer who understands JS up to the point where it is converted into bytecode and executed on the hardware level. I follow his work quite closely. He is increasingly opening my eyes to how JS really works.

By the way, if you replace all let/const in your code with var (if you thought it was outdated, you are very wrong), you can significantly optimize the framework's performance. The reason is that let/const uses certain checks that are completely absent in var. It also turned out that the concept of var is very logical, and let/const is overused, which spoiled the expressiveness of the language.

But again, modern runtimes can optimize such things if you don't interfere with them, and it's very easy to interfere with them)))

File: src/mod.ts [hunk 1]
```diff
@@ -32,18 +34,13 @@ export type EmojiFlavor<C extends Context = Context> = C & {
...
-function withEmoji(string: TemplateStringsArray, ...emojis: EmojiName[]) {
-    return string.reduce((acc, str, index) => {
-        // The number of elements in `string` is 1 more than in `emojis`, therefore
-        // we will be looking for an `undefined` emoji, which does not make sense,
-        // so we exit the loop early.
-        if (index === string.length - 1) return acc + str;
-        const emoji = getEmoji(emojis[index]);
-        return acc + str + (typeof emoji === "string" ? emoji : emoji.emoji);
+const getEmoji = (name: EmojiName) => emojis[name as keyof EmojiList] || name;
+const withEmoji = (text: TemplateStringsArray, ...emojis: EmojiName[]) => {
```
Comment (dcdunkan) [inline src/mod.ts]: I'd like it if there were spaces between function definitions. I think it'll make the code slightly more readable.

**Review by dcdunkan — APPROVED** [review]
LGTM.

**Comment by dcdunkan** [comment]
@KnorpelSenf, I don't see any breaking changes in this. What should we do about the next package versioning, given that the emoji list have been corrected and updated? I think we can make it a minor release, because now we at least leave the incorrect emoji names as it is.

File: src/mod.ts [hunk 2]
```diff
@@ -32,18 +34,13 @@ export type EmojiFlavor<C extends Context = Context> = C & {
...
 
-function withEmoji(string: TemplateStringsArray, ...emojis: EmojiName[]) {
-    return string.reduce((acc, str, index) => {
-        // The number of elements in `string` is 1 more than in `emojis`, therefore
-        // we will be looking for an `undefined` emoji, which does not make sense,
-        // so we exit the loop early.
-        if (index === string.length - 1) return acc + str;
-        const emoji = getEmoji(emojis[index]);
-        return acc + str + (typeof emoji === "string" ? emoji : emoji.emoji);
+const getEmoji = (name: EmojiName) => emojis[name as keyof EmojiList] || name;
```
Comment (KnorpelSenf) [inline src/mod.ts]: Semantically, we want to replace absent values, not empty ones.
```suggestion
const getEmoji = (name: EmojiName) => emojis[name as keyof EmojiList] ?? name;
```

File: src/mod.ts [hunk 3]
```diff
@@ -52,10 +49,5 @@ export function emojiParser<C extends EmojiFlavor>() {
...
     };
-}
-
-export function emoji(name: EmojiName): string {
-    const emoji = getEmoji(name);
-    // `emoji` will be `name` if an emoji with that name doesn't exist.
-    return typeof emoji === "string" ? emoji : emoji.emoji;
-}
+};
+export const emoji = (name: EmojiName) => getEmoji(name);
```
Comment (KnorpelSenf) [inline src/mod.ts]: At this point we should just rename and export `getEmoji` directly.

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]
In general, grammY is used in networking contexts. The high performance penalties come from async ops such as network calls or file reads. The highest performance gains can be achieved by performing async ops concurrently rather than sequentially. We are solving an IO-bound task, not a CPU-bound one, and the performed CPU work is negligible. By far the most expensive CPU operation (commonly approx. 80 % of CPU work) is to parse and serialise requests and responses when communicating with the Bot API. For all of these reasons, looking at the CPU cost of function objects vs. lambdas is the wrong thing to measure, and optimising this code is (interesting and fun) but a waste of time and bad priorisation from an engineering perspective.

We care a lot about readability, and writing code that other people can understand trivially is of core importance. Improving an edge case of CPU perf even by 1 % (which is an absurdly high assumption) would not justify the deterioration in readability.

What we can indeed consider is to store the emoji data in a JSON file, as this file is half a megabyte in size, so we might be able to improve startup time by a millisecond or so, which could be nice. I'm not saying it's relevant, but eh, why not. The main challenge here is that we are a hybrid plugin, so we can't simply `require` the file. When running `deno cache`, we would ideally want the file to be downloaded, and loading JSON at runtime might prevent that, hence making startup time substantially worse (emoji data would have to be downloaded from the web on every startup). For Node, we could add a special case that loads the file from the local `node_modules` folder.

Apart from all of that, I think the most significant performance improvement is to only load partial data. We could either
- let the user specify a custom emoji data source so they can strip down the file to only the data they use, or
- do implicit, opaque and evil dark magic where we cache the used emoji in a JSON file locally somewhere so we only have to load this subset upon the next start, and then fall back to loading the full dataset if an unknown emoji is requested.

**Comment by KnorpelSenf** [comment]
Then again, the current setup is very simple yet very reliable, and messing around with it to save a few ms at best once during startup might not be worth the headache. This is something for @dcdunkan to decide :)

File: src/mod.ts [hunk 4]
```diff
@@ -1,6 +1,8 @@
 import { Context, NextFunction } from "./deps.deno.ts";
-import getEmoji from "./emoji.ts";
-import type { EmojiName } from "./emoji.ts";
+import emojis from "./emojidata.ts";
+
+type EmojiList = typeof emojis;
+type EmojiName = keyof EmojiList | (string & Record<never, never>);
```
Comment (KnorpelSenf) [inline src/mod.ts:5]: This is clever, do you feel like opening a pull request for the core repo that switches to `Record<never, never>` for `StringWithSuggestions`? That way, we can enable linting for the respective lines.

**Review by KnorpelSenf — APPROVED** [review]
LGTM, thanks!

File: src/mod.ts (hunk 4 above)
Comment (niusia-ua) [reply src/mod.ts:5]: I don't get it. What does this improve? In the grammY core, do you also need to use `string & Record<never, never>` somewhere?

Comment (KnorpelSenf) [reply src/mod.ts:5]: https://github.com/grammyjs/grammY/blob/8acfe06252fb7d7b8a1f089b2446c5f117f6748c/src/context.ts#L32

**Comment by dcdunkan** [comment]
> Then again, the current setup is very simple yet very reliable, and messing around with it to save a few ms at best once during startup might not be worth the headache. This is something for @dcdunkan to decide :)

I agree with you after reading your comment. I think the setup we currently have in this PQ is great and simple enough.

**Review by dcdunkan — APPROVED** [review]
Thank you so much!

**Pull request merged**